File: python/atelier_core/billing/formatter.py
"""
Billing-side formatter.

Shadow pair: `atelier_core.reporting.formatter` exports a class with the same
name and an unrelated method set. Consumers alias at the import site.
"""

from __future__ import annotations

from ..money import Money


class Formatter:
    def __init__(self, currency: str = "EUR") -> None:
        self.currency = currency

    def money(self, amount: Money, currency: str = "") -> str:
        return f"{amount} {currency or self.currency}"

    def line(self, quantity: int, description: str, amount: Money) -> str:
        return f"{quantity} x {description} = {self.money(amount)}"

File: go/core/billing/formatter.go
// Package billing is one half of the SHADOW PAIR: it exports a type named
// Formatter, and so does core/reporting. The two are unrelated; consumers
// alias one or both at the import site.
package billing

import (
	"fmt"

	"atelier.example/lane/core/money"
)

// Formatter renders money. The reporting half renders statuses.
type Formatter struct {
	Currency string
}

// NewFormatter defaults to euros.
func NewFormatter(currency string) *Formatter {
	if currency == "" {
		currency = "EUR"
	}
	return &Formatter{Currency: currency}
}

// Money renders "349.00 EUR".
func (f *Formatter) Money(amount money.Money) string {
	return fmt.Sprintf("%s %s", amount, f.Currency)
}

// Line renders "2 x Battery 55Wh = 178.00 EUR".
func (f *Formatter) Line(quantity int, description string, amount money.Money) string {
	return fmt.Sprintf("%d x %s = %s", quantity, description, f.Money(amount))
}
